ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosChangeFeedBasedReplicationFeed.java

```diff
@@ -0,0 +1,251 @@
+     * this means that the cached set of fetches entries hasn't been consumed within the invalidation duration.
+     * @return true if this entry is expired. false otherwise.
+     */
+    boolean isExpired() {
+      return creationTimestamp < System.currentTimeMillis() - CACHE_VALID_DURATION_IN_MS;
+    }
+  }
+
+  // change feed cache by cache session id
+  private final ConcurrentHashMap<String, ChangeFeedCacheEntry> changeFeedCache;
+  private final int defaultCacheSize;
+  private final CosmosDataAccessor cosmosDataAccessor;
+  private final AzureMetrics azureMetrics;
+  private final static long CACHE_VALID_DURATION_IN_MS = 60 * 60 * 1000; //1 hour
+
+  /**
+   * Constructor to create a {@link CosmosChangeFeedBasedReplicationFeed} object.
+   * @param cosmosDataAccessor {@link CosmosDataAccessor} object.
+   * @param azureMetrics{@link {@link AzureMetrics} object.
+   */
+  public CosmosChangeFeedBasedReplicationFeed(CosmosDataAccessor cosmosDataAccessor, AzureMetrics azureMetrics) {
+    this.defaultCacheSize = AzureCloudDestination.getFindSinceQueryLimit();
+    changeFeedCache = new ConcurrentHashMap<>();
+    this.cosmosDataAccessor = cosmosDataAccessor;
+    this.azureMetrics = azureMetrics;
+    // schedule periodic invalidation of cache
+    Utils.newScheduler(1, false)
+        .scheduleAtFixedRate(() -> changeFeedCache.entrySet().removeIf(entry -> entry.getValue().isExpired()),
+            CACHE_VALID_DURATION_IN_MS, CACHE_VALID_DURATION_IN_MS, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Get next set of change feed entries for the specified partition, after the {@code curFindToken}.
+   * The number of entries is capped by maxEntriesSize.
+   * This method creates a cache for change feed entries. If the {@code curFindToken} is not valid,
+   * or if all the items in the cache are consumed, then it queries Cosmos for new entries.
+   * @param curFindToken {@link FindToken} after which the next entries have to be returned.
+   * @param maxTotalSizeOfEntries maximum size of all the blobs returned.
+   * @param partitionPath Partition for which change feed entries have to be returned.
+   * @return {@link FindResult} instance that contains updated {@link FindToken} object which can act as a bookmark for
+   * subsequent requests, and {@link List} of {@link CloudBlobMetadata} entries.
+   * @throws {@link DocumentClientException}.
+   */
+  @Override
+  public FindResult getNextEntriesAndUpdatedToken(FindToken curFindToken, long maxTotalSizeOfEntries,
+      String partitionPath) throws DocumentClientException {
+    List<CloudBlobMetadata> nextEntries = new ArrayList<>();
+    CosmosChangeFeedFindToken cosmosChangeFeedFindToken = (CosmosChangeFeedFindToken) curFindToken;
+    int index = cosmosChangeFeedFindToken.getIndex();
+    String cacheSesionId = cosmosChangeFeedFindToken.getCacheSessionId();
+    if (!changeFeedCache.containsKey(cacheSesionId) || !isCacheValid(partitionPath, cosmosChangeFeedFindToken)) {
```

Comment:
I would recommend trying to minimize the number of `changeFeedCache.get` calls since the background thread can change the state of the map from underneath.
Assuming populateChangeFeedCache returns a `CacheEntry` object and `isCacheValid` takes in the `CacheEntry` instead of getting again, you could do something like:
```java
CacheEntry cacheEntry = changeFeedCache.get(sessionId);
if (cacheEntry == null || !isCacheValid(cacheEntry, partitionPath, cosmosChangeFeedFindToken)) {
  cacheEntry = populateChangeFeed(...);
  ...
}
...
    } else {
        cacheEntry = populateChangeFeedCache(partitionPath, cosmosChangeFeedFindToken.getEndContinuationToken(),
            cosmosChangeFeedFindToken.getCacheSessionId());
```